core/src/main/java/com/bettercloud/bigtable/orm/BigTableEntityDao.java

```diff
@@ -125,48 +127,97 @@
 
-                        if (bytes.length > 0) {
-                            value = objectMapper.readValue(bytes, column.getTypeReference());
-                        } else {
-                            value = null;
-                        }
-                    } else {
-                        value = null;
-                    }
+            final Cell cell = result.getColumnLatestCell(family, qualifier);
 
-                    delegate.setColumnValue(column, value);
+            final Object value;
 
-                    if (column.isVersioned()) {
-                        final Long timestamp = Optional.ofNullable(cell)
-                                .map(Cell::getTimestamp)
-                                .orElse(null);
+            if (cell != null) {
+                final byte[] bytes = cell.getValueArray();
 
-                        delegate.setColumnTimestamp(column, timestamp);
-                    }
+                if (bytes.length > 0) {
+                    value = objectMapper.readValue(bytes, column.getTypeReference());
+                } else {
+                    value = null;
                 }
+            } else {
+                value = null;
+            }
+
+            delegate.setColumnValue(column, value);
 
-                entitiesByKey.put(entry.getKey(), entity);
+            if (column.isVersioned()) {
+                final Long timestamp = Optional.ofNullable(cell).map(Cell::getTimestamp).orElse(null);
+
+                delegate.setColumnTimestamp(column, timestamp);
             }
         }
 
+        return entity;
+    }
+
+    /**
+     * If rows exists that match the scan, then the resulting Map will contain the key and a value of type {@link T} for each matching result.
+     *
+     * If the scan does not match any rows, then the resulting Map will be empty.
+     *
+     * The returned objects will contain data for each configured column only if the corresponding column actually
+     * contained data. If data is found, then it is deserialized using the configured
+     * {@link com.fasterxml.jackson.core.type.TypeReference} for that column. If no data is found, then the
+     * corresponding value for that column is set to null.
+     *
+     * @param scan The {@link Scan} that should be used for querying BigTable.
+     * @param <K> The type of the keys used to retrieve the rows.
+     * @return A Map containing pairs of keys and their corresponding values.
+     * @throws IOException when an error occurs while communicating with BigTable
+     */
+    @Override
+    public <K extends Key<T>> Map<K, T> scan(final Scan scan) throws IOException {
+        Objects.requireNonNull(scan);
+
+        final Map<K, T> entitiesByKey = new HashMap<>();
+
+        final ResultScanner scanner = table.getScanner(scan);
+        for (Result result : scanner) {
+            final K key = getRowKey(result);
+            final T entity = buildEntity(result);
```

Comment:
Scans present a tricky issue for deserialization. Consider the following:

The `@Table` configuration contains 2 `@Entity` definitions, A and B. Both A and B declare a `@Column(family = "f", qualifier = "q")`. However, A declares that column to contain `String`s, and B declares that column to contain `Float`s.

If we use `Dao<A>`, then the `scan` method will find rows that were originally written using `Dao<B>`, therefore attempting to deserialize fields as Strings, which were meant to be Floats. This will cause the ObjectMapper to throw IOExceptions on line 166.

There are a few options:

1) Trust that the user knows what they're doing (I tend not to, otherwise this library wouldn't exist)
2) Catch IOExceptions and simply don't return fields that failed to deserialize (super expensive on failures, and the possibility of something not being returned that should have been valid, leaving the user clueless as to what's going on)
3) Return a `Map<K, SomeResult<V>>`, where `SomeResult` contains the `byte[]` value, the deserialized value (on success), and the caught exception (on failure).
4) Don't implement scans within this library, and make the user handle "advanced" use-cases. Maybe split deserialization into its own interface/classes so that the user can access those separately from the `Dao` itself, and just have the `Dao` use it (this allows you to continue using the Scan objects from the userspace that I was opposed to in the other comment)

I'm gonna let BetterCloud make the final call, but it's important to consider the implications here.

Reply:
I think you are right and it doesn't make sense to put the "advanced" cases into this library.  I will close this PR.  Thanks for the feedback.